### PR TITLE
PCHR-2302: Cannot save assignment with more then 20 documents

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
Currently, when trying to save the assignments in T&A with more than 20 documents, the assignment can't be saved due to following error:
```
POST http://176.58.123.15:8300/civicrm/ajax/rest?action=create_multiple&entity=D…2017-06-06T21:00:00.000Z%22,%22target_contact_id%22:%5B%222%22%5D%7D%5D%7D 414 (Request-URI Too Large)
```
The error was caused, because when saving the assignment with many documents, the data is sent through url adding a query string.

Similar fix is already applied in PR #188 for saving multiple tasks in Assignment.

This issue is fixed making amendments `$resource` in Document Service, to send document data in request body rather than in url as query string.

#### Assignment with more than 20 documents cannot be saved raising the error.
request ui too large

#### Assignment with more than 20 documents can now be saved.
2302_final-min

## Technical Details
1. Modify `$resource` in Document service and override `save()` call to send data in request body.
```js
services.factory('Document', ['$resource',  '$httpParamSerializer', 'config', '$log', function ($resource, $httpParamSerializer, config, $log) {
    $log.debug('Service: Document');

    return $resource(config.url.REST, {
      'action': 'get',
      'entity': 'Document',
      'sequential': 1,
      'debug': config.DEBUG
    }, {
      save: {
        method: "POST",
        isArray: false,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
        },
        transformRequest: $httpParamSerializer
      }
    });
  }]);
```
2. Amend `assign()` function to adopt  changes in `$resource`
```js
assign: function (documentArr, assignmentId) {
          ...
          Document.save({ action: 'copy_to_assignment' }, {
            json: {
              id: documentArr,
              case_id: assignmentId
            } || {}
          }, function (data) {
            ...
        }
```
3. Split `getOptions() call to separate functions for getting document statuses and document types respectively. 
```js
getOptions: function() {
          var deferred = $q.defer();

          $q.all({
            documentType: this.getDocumentTypes(),
            documentStatus: this.getDocumentStatus()
          }).then(function(options) {
            deferred.resolve(options);
          });

          return deferred.promise;
        },
        getDocumentStatus: function() {
          var deferredDocumentStatus = $q.defer();
          var documentStatus = {
              arr: [],
              obj: {}
            };

          Document.get({
            action: 'getoptions',
            json: {
              'field': 'status_id'
            }
          }, function(data) {
            _.each(data.values, function(option) {
              documentStatus.arr.push({
                key: option.key.toString(),
                value: option.value
              });

              documentStatus.obj[option.key] = option.value;
            });

            deferredDocumentStatus.resolve(documentStatus);
          });

          return deferredDocumentStatus.promise;
        },
        getDocumentTypes: function() {
          var deferredDocumentType = $q.defer();
          var documentType = {
              arr: [],
              obj: {}
            };

          Document.get({
            action: 'getoptions',
            json: {
              'field': 'activity_type_id',
              'options': {
                'limit': 0
              }
            }
          }, function(data) {
            _.each(data.values, function(option) {
              documentType.arr.push({
                key: option.key,
                value: option.value
              });

              documentType.obj[option.key] = option.value;
            });

            deferredDocumentType.resolve(documentType);
          });

          return deferredDocumentType.promise;
        },
```
4. Amend `saveMultiple()` to adopt changes in `$resource` in Document Service.
```
saveMultiple: function (documentArr) {
     ....
      Document.save({ action: 'create_multiple'}, {
            json: {
              document: documentArr
            } || {}
          }, function (data) {
             ....
          }
})
```
5. Amend `save()` to adopt changes in `$resource`
```js
save: function (document) {
         Document.save({ action: 'create' }, {
            json: params || {}
          }, function (data) {
              ...
          }
});
```
6. Amend `sendReminder()` to adopt changes in `$resource`
```js
sendReminder: function (documentId, notes) {
        Document.save({ action: 'sendreminder' }, {
            json: {
              activity_id: documentId,
              notes: notes || ''
            } || {}
          }, function (data) {
              ...
          }
})
```
## Comments
- Added related test cases for the changes introduced in this PR

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js [hunk 1]
```diff
@@ -78,67 +85,74 @@ define([
 
           return deferred.promise;
         },
-        getOptions: function () {
+        getOptions: function() {
           var deferred = $q.defer();
-          var deferredDocumentType = $q.defer();
+
+          $q.all({
```
Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: I tried to implement your solution, but doing so, the promise does not get resolved.

Comment (bmnepali) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: Updated!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/* globals angular */
+/* globals angular, _ */
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: sorry, just caught this. `angular` and `_` cannot be global variables. They come from the `common/angular` and `common/lodash` module, so just inject those

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js [hunk 3]
```diff
@@ -79,13 +86,15 @@ define([
           return deferred.promise;
         },
         getOptions: function () {
-          var deferred = $q.defer();
-          var deferredDocumentType = $q.defer();
+          return $q.all({
+              documentType: this.getDocumentTypes(),
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: wrong indentation

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js [hunk 4]
```diff
@@ -79,13 +86,15 @@ define([
...
+              documentStatus: this.getDocumentStatus()
+            }).then(function (options) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: if you're just returning `options` without doing anything with it, then you can skip the `then` altogether

**Pull request merged**